VnodeCacheDesign.md

```diff
@@ -0,0 +1,55 @@
+# Vnode Cache Design
+
+## Goals
+
+- Reduce the time required to determine if a file is inside of a virtualization root
+- Reduce the time required to find a file's virtualization root
+- **TBD:** Eliminate the use of file flags to determine if a file is inside of a root
```

Comment:
I think getting rid of the `IsInVirtualizationRoot` flag should be a primary goal of this design. That flag exists only as a way to short-circuit our logic for paths outside of any virtualization root, but it is flaky for several reasons. The main benefit of this cache is that we can answer the question cheaply and reliably.

Reply:
> I think getting rid of the IsInVirtualizationRoot flag should be a primary goal of this design. That flag exists only as a way to short-circuit our logic for paths outside of any virtualization root, but it is flaky for several reasons. The main benefit of this cache is that we can answer the question cheaply and reliably.

Agree on all points.

I think the open question is how frequently we'll miss in our cache.  If the miss % is too high then we'll end up slowing everything down walking up the vnode hierarchy.

Reply:
> Additionally, are we interested in removing the need for the 'empty' flag too, or are we considering that to be out of scope for now?

@pmj I think it would be great to eliminate the "empty" flag too, but I think that's out of scope for this first iteration of the vnode cache.

Out of curiosity, did you have any idea(s) in mind for how the vnode cache could address this?  

Even if it's out of scope right in the initial implementation I'd like to avoid going down a path that makes it harder to add later.

Reply:
We can persist the empty state in the xattr, and cache that information in the vnode cache as a simple flag when we end up reading the xattr. We can invalidate the cache when we receive an xattr write check in the vnode listener, and also after receiving a success response for a hydrate file provider message.

Reply:
@pmj thanks for the suggestion!  I'll include the details you've suggested in the doc

Reply:
FWIW, I believe @sanoursa did not like this approach when I first suggested it back in ~June, but then he was pretty lukewarm on the vnode cache idea in general. 😃 If I remember correctly, the core concern was that we'd be tracking state in more than one place and the possibility that these different versions of the same information might go out of sync.